README.rst
extension-helpers
=================

.. image:: https://travis-ci.org/astropy/extension-helpers.svg
  :target: https://travis-ci.org/astropy/extension-helpers

.. image:: https://ci.appveyor.com/api/projects/status/rt9161t9mhx02xp7/branch/master?svg=true
  :target: https://ci.appveyor.com/project/Astropy/extension-helpers

.. image:: https://codecov.io/gh/astropy/extension-helpers/branch/master/graph/badge.svg
  :target: https://codecov.io/gh/astropy/extension-helpers

The **extension-helpers** package includes convenience helpers to assist with
building Python packages with compiled C/Cython extensions. It is developed by
the Astropy project but is intended to be general and usable by any Python
package.

For more information, see the documentation at http://extension-helpers.readthedocs.io


CHANGES.rst
astropy-helpers Changelog
*************************

4.1 (unreleased)
----------------

- No changes yet.


4.0.2 (unreleased)
------------------

- Fixed leaked temporary directory in ``check_openmp_support``. [#511]


4.0.1 (2019-12-16)
------------------

- Fix pip installation. [#506]


4.0 (2019-12-13)
----------------

- Changed minimum required Python version to 3.6. [#498]

- Changed minimum required Sphinx version to 1.7. [#498]

- Added a --parallel option for build_docs. [#498]

- Fix compatibility with setuptools 42.x. [#504]


3.2.2 (2019-10-25)
------------------

- Correctly handle main package directory inside namespace package. [#486]


3.2.1 (2019-06-13)
------------------

- Reverting issuing deprecation warning for the ``build_sphinx`` command. [#482]

- Make sure that all data files get included in tar file releases. [#485]


3.2 (2019-05-29)
----------------

- Make sure that ``[options.package_data]`` in setup.cfg is taken into account
  when collecting package data. [#453]

- Simplified the code for the custom build_ext command. [#446]

- Avoid importing the astropy package when trying to get the test command
  when testing astropy itself. [#450]

- Avoid importing whole package when trying to get version information. Note
  that this has also introduced a small API change - ``cython_version`` and
  ``compiler`` can no longer be imported from the ``package.version`` module
  generated by astropy-helpers. Instead, you can import these from
  ``package.cython_version`` and ``package.compiler_version`` respectively. [#442]

- Make it possible to call ``generate_version_py`` and ``register_commands``
  without any arguments, which causes information to be read in from the
  ``setup.cfg`` file. [#440]

- Simplified setup.py and moved most of the configuration to setup.cfg. [#445]

- Add a new ``astropy_helpers.setup_helpers.setup`` function that does all
  the default boilerplate in typical ``setup.py`` files that use
  astropy-helpers. [#443]

- Remove ``deprecated``, ``deprecated_attribute``, and ``minversion`` from
  ``astropy_helpers.utils``. [#447]

- Updated minimum required version of setuptools to 30.3.0. [#440]

- Remove functionality to adjust compilers if a broken compiler is detected.
  This is not useful anymore as only a single compiler was previously patched
  (now unlikely to be used) and this was only to fix a compilation issue in the
  core astropy package. [#421]

- ``sphinx-astropy`` is now a required dependency to build the docs, the
  machinery to install it as eggs have been removed. [#474]


3.1.1 (2019-02-22)
------------------

- Moved documentation from README to Sphinx. [#444]

- Fixed broken OpenMP detection when building with ``-coverage``. [#434]


3.1 (2018-12-04)
----------------

- Added extensive documentation about astropy-helpers to the README.rst file. [#416]

- Fixed the compatibility of the build_docs command with Sphinx 1.8 and above. [#413]

- Removing deprecated test_helpers.py file. [#369]

- Removing ez_setup.py file and requiring setuptools 1.0 or later. [#384]

- Remove all sphinx components from ``astropy-helpers``. These are now replaced
  by the ``sphinx-astropy`` package in conjunction with the ``astropy-theme-sphinx``,
  ``sphinx-automodapi``, and ``numpydoc`` packages. [#368]

- openmp_helpers.py: Make add_openmp_flags_if_available() work for clang.
  The necessary include, library, and runtime paths now get added to the C test code
  used to determine if openmp works.
  Autogenerator utility added ``openmp_enabled.is_openmp_enabled()``
  which can be called post build to determine state of OpenMP support.
  [#382]

- Add version_info tuple to autogenerated version.py. Allows for simple
  version checking, i.e. version_info > (2,0,1). [#385]


3.0.2 (2018-06-01)
------------------

- Nothing changed.


3.0.1 (2018-02-22)
------------------

- Nothing changed.


3.0 (2018-02-09)
----------------

- Removing Python 2 support, including 2to3. Packages wishing to keep Python
  2 support should NOT update to this version. [#340]

- Removing deprecated _test_compat making astropy a hard dependency for
  packages wishing to use the astropy tests machinery. [#314]

- Removing unused 'register' command since packages should be uploaded
  with twine and get registered automatically. [#332]


2.0.11 (2019-10-25)
-------------------

- Fixed deprecation warning in sphinx theme. [#493]

- Fixed an issue that caused pytest to crash if it tried to collect
  tests. [#488]


2.0.10 (2019-05-29)
-------------------

- Removed ``tocdepthfix`` sphinx extension that worked around a big in
  Sphinx that has been long fixed. [#475]

- Allow Python dev versions to pass the python version check. [#476]

- Updated bundled version of sphinx-automodapi to v0.11. [#478]


2.0.9 (2019-02-22)
------------------

- Updated bundled version of sphinx-automodapi to v0.10. [#439]

- Updated bundled sphinx extensions version to sphinx-astropy v1.1.1. [#454]

- Include package name in error message for Python version in
  ``ah_bootstrap.py``. [#441]


2.0.8 (2018-12-04)
------------------

- Fixed compatibility with Sphinx 1.8+. [#428]

- Fixed error that occurs when installing a package in an environment where
  ``numpy`` is not already installed. [#404]

- Updated bundled version of sphinx-automodapi to v0.9. [#422]

- Updated bundled version of numpydoc to v0.8.0. [#423]


2.0.7 (2018-06-01)
------------------

- Removing ez_setup.py file and requiring setuptools 1.0 or later. [#384]


2.0.6 (2018-02-24)
------------------

- Avoid deprecation warning due to ``exclude=`` keyword in ``setup.py``. [#379]


2.0.5 (2018-02-22)
------------------

- Fix segmentation faults that occurred when the astropy-helpers submodule
  was first initialized in packages that also contained Cython code. [#375]


2.0.4 (2018-02-09)
------------------

- Support dotted package names as namespace packages in generate_version_py.
  [#370]


astropy_helpers/version.py
# Autogenerated by Astropy-affiliated package astropy_helpers's setup.py on 2019-11-13 09:29:46 UTC
import datetime


import locale
import os
import subprocess
import warnings

__all__ = ['get_git_devstr']


def _decode_stdio(stream):
    try:
        stdio_encoding = locale.getdefaultlocale()[1] or 'utf-8'
    except ValueError:
        stdio_encoding = 'utf-8'

    try:
        text = stream.decode(stdio_encoding)
    except UnicodeDecodeError:
        # Final fallback
        text = stream.decode('latin1')

    return text


def update_git_devstr(version, path=None):
    """
    Updates the git revision string if and only if the path is being imported
    directly from a git working copy.  This ensures that the revision number in
    the version string is accurate.
    """

    try:
        # Quick way to determine if we're in git or not - returns '' if not
        devstr = get_git_devstr(sha=True, show_warning=False, path=path)
    except OSError:
        return version

    if not devstr:
        # Probably not in git so just pass silently
        return version

    if 'dev' in version:  # update to the current git revision
        version_base = version.split('.dev', 1)[0]
        devstr = get_git_devstr(sha=False, show_warning=False, path=path)

        return version_base + '.dev' + devstr
    else:
        # otherwise it's already the true/release version
        return version


def get_git_devstr(sha=False, show_warning=True, path=None):
    """
    Determines the number of revisions in this repository.

    Parameters
    ----------
    sha : bool
        If True, the full SHA1 hash will be returned. Otherwise, the total
        count of commits in the repository will be used as a "revision
        number".

    show_warning : bool
        If True, issue a warning if git returns an error code, otherwise errors
        pass silently.

    path : str or None
        If a string, specifies the directory to look in to find the git
        repository.  If `None`, the current working directory is used, and must
        be the root of the git repository.
        If given a filename it uses the directory containing that file.

    Returns
    -------
    devversion : str
        Either a string with the revision number (if `sha` is False), the
        SHA1 hash of the current commit (if `sha` is True), or an empty string
        if git version info could not be identified.

    """

    if path is None:
        path = os.getcwd()

    if not os.path.isdir(path):
        path = os.path.abspath(os.path.dirname(path))

    if sha:
        # Faster for getting just the hash of HEAD
        cmd = ['rev-parse', 'HEAD']
    else:
        cmd = ['rev-list', '--count', 'HEAD']

    def run_git(cmd):
        try:
            p = subprocess.Popen(['git'] + cmd, cwd=path,
                                 stdout=subprocess.PIPE,
                                 stderr=subprocess.PIPE,
                                 stdin=subprocess.PIPE)
            stdout, stderr = p.communicate()
        except OSError as e:
            if show_warning:
                warnings.warn('Error running git: ' + str(e))
            return (None, b'', b'')

        if p.returncode == 128:
            if show_warning:
                warnings.warn('No git repository present at {0!r}! Using '
                              'default dev version.'.format(path))
            return (p.returncode, b'', b'')
        if p.returncode == 129:
            if show_warning:
                warnings.warn('Your git looks old (does it support {0}?); '
                              'consider upgrading to v1.7.2 or '
                              'later.'.format(cmd[0]))
            return (p.returncode, stdout, stderr)
        elif p.returncode != 0:
            if show_warning:
                warnings.warn('Git failed while determining revision '
                              'count: {0}'.format(_decode_stdio(stderr)))
            return (p.returncode, stdout, stderr)

        return p.returncode, stdout, stderr

    returncode, stdout, stderr = run_git(cmd)

    if not sha and returncode == 128:
        # git returns 128 if the command is not run from within a git
        # repository tree. In this case, a warning is produced above but we
        # return the default dev version of '0'.
        return '0'
    elif not sha and returncode == 129:
        # git returns 129 if a command option failed to parse; in
        # particular this could happen in git versions older than 1.7.2
        # where the --count option is not supported
        # Also use --abbrev-commit and --abbrev=0 to display the minimum
        # number of characters needed per-commit (rather than the full hash)
        cmd = ['rev-list', '--abbrev-commit', '--abbrev=0', 'HEAD']
        returncode, stdout, stderr = run_git(cmd)
        # Fall back on the old method of getting all revisions and counting
        # the lines
        if returncode == 0:
            return str(stdout.count(b'\n'))
        else:
            return ''
    elif sha:
        return _decode_stdio(stdout)[:40]
    else:
        return _decode_stdio(stdout).strip()


# This function is tested but it is only ever executed within a subprocess when
# creating a fake package, so it doesn't get picked up by coverage metrics.
def _get_repo_path(pathname, levels=None):  # pragma: no cover
    """
    Given a file or directory name, determine the root of the git repository
    this path is under.  If given, this won't look any higher than ``levels``
    (that is, if ``levels=0`` then the given path must be the root of the git
    repository and is returned if so.

    Returns `None` if the given path could not be determined to belong to a git
    repo.
    """

    if os.path.isfile(pathname):
        current_dir = os.path.abspath(os.path.dirname(pathname))
    elif os.path.isdir(pathname):
        current_dir = os.path.abspath(pathname)
    else:
        return None

    current_level = 0

    while levels is None or current_level <= levels:
        if os.path.exists(os.path.join(current_dir, '.git')):
            return current_dir

        current_level += 1
        if current_dir == os.path.dirname(current_dir):
            break

        current_dir = os.path.dirname(current_dir)

    return None


_packagename = "astropy_helpers"
_last_generated_version = "4.0.dev1188"
_last_githash = "a1b4e02d7b483991749bf7b2358900b18768837a"

# Determine where the source code for this module
# lives.  If __file__ is not a filesystem path then
# it is assumed not to live in a git repo at all.
if _get_repo_path(__file__, levels=len(_packagename.split('.'))):
    version = update_git_devstr(_last_generated_version, path=__file__)
    githash = get_git_devstr(sha=True, show_warning=False,
                             path=__file__) or _last_githash
else:
    # The file does not appear to live in a git repo so don't bother
    # invoking git
    version = _last_generated_version
    githash = _last_githash


major = 4
minor = 0
bugfix = 0

version_info = (major, minor, bugfix)

release = False
timestamp = datetime.datetime(2019, 11, 13, 9, 29, 46)
debug = False

astropy_helpers_version = "4.0.dev1188"
